net/modemmanager/files/modemmanager.proto

```diff
@@ -404,6 +406,33 @@ proto_modemmanager_setup() {
 		return 1
 	}
 
+	# set the RAT
+	if [ -n "${rat}" ]; then
+		if [ "${rat}" == "automatic" ]; then
+			echo "setting modem RAT to automatic mode"
+			if [ "$(mmcli --modem="${device}" | grep supported | grep 5gnr)" ]; then
```

Comment:
This is not completely right I'm afraid. The "5gnr", "lte" and "gsm-umts" strings being grepped here are reporting "capabilities", not "modes", it is not exactly the same in the context of ModemManager.

For "automatic", the setting to use is simple: `--set-allowed-modes=any --set-preferred-mode=none`. This setting is built-in ModemManager and will try to automatically map to a setting that means "auto", regardless of the proto in use. There is no need to guess anything here.

Then, for the "5G" rat selection, the modem is assuming that 4G is also supported, while that may not be true. A 5GNR-only device doesn't have to support 4G.

And then, for the "4G" and "3G" rat selections, the user is selecting **only** those rats. E.g. a 4G/3G/2G modem selecting 4G would be selecting 4G-only, which may not be completely desired.

And 2G is completely ignored in the patch, while it shouldn't. There are a lot of 4G+2G modems (without 3G) as well.

My suggestion here would be:
 * First, let the proto handler receive exactly which allowed modes and preferred modes the user wants. So not a "rat" setting, but explicitly two settings saying "allowed-modes" and "preferred-modes", and pass those settings to ModemManager with the `--set-preferred-mode` and `--set-allowed-modes` commands without doing any explicit validation. The validation of what is supported or not is done by the modem itself anyway, so if a user is manually configuring those in the config file, the user should be aware of which settings are supported.
 * Second, in the luci-proto-modemmanager package, we can expose a list of which are the allowed/preferred combinations the user can choose, as reported in the `mmcli -m $device` output. Any of those combinations listed will be supported by the modem (and any combination NOT in the list will NOT be supported).

I'm afraid that this patch has made a lot of assumptions about a very specific modem model being used, I think it can be improved to make it more generic.